### PR TITLE
Hypothesis for objectives

### PR DESCRIPTION
**Pull request by AdrianSosic**
Adds the hypothesis strategies for our objective classes.

### CODE REVIEW

**Comment by AdrianSosic** [comment]
Hi @AVHopp @Scienfitz. This is a short PR on top of the open #188 PR to add the strategies. Please wait with the review until the other one is merged.

**Comment by AdrianSosic** [comment]
@AVHopp, @Scienfitz: Ok, now this one is also ready for review

File: baybe/objectives/base.py
Comment (Scienfitz) [inline baybe/objectives/base.py:21]: did this cause an issue with hypothesis or is is just for consistency?

File: tests/hypothesis_strategies/targets.py
Comment (Scienfitz) [inline tests/hypothesis_strategies/targets.py:15]: i think we have not been super accurate with type annotations in these strategies

eg here im not even sure whats the intended parameter type? If its a single given interval, this parameter should prob even be renamed to `interval` not `intervals`, but maybe you intend this to be a strategy?

File: tests/hypothesis_strategies/objectives.py
Comment (Scienfitz) [inline tests/hypothesis_strategies/objectives.py:27]: do we explicitly want to allow a (nonsensical) weight of 0.0 ? conceptually this target then should simply be ignored, but maybe there could be numerical problems, e.g. in the geom_mean or mean utility functions?

**Review by Scienfitz — APPROVED** [review]
looks goodcci , minor comments

File: baybe/objectives/desirability.py
Comment (AVHopp) [inline baybe/objectives/desirability.py:114]: If this property just returns the private `_targets`, what is the point of making that field private?

File: baybe/objectives/desirability.py
Comment (AVHopp) [inline baybe/objectives/desirability.py:147]: You need to change the comment, since this is also the point where normalization happens, which is not obvious imo.

File: baybe/objectives/base.py
Comment (AdrianSosic) [reply baybe/objectives/base.py:21]: No, not with hypothesis but because I added a cached property to the class (had to refactor the weight normalization issue due to the reason I briefly mentioned in one of our meetings, i.e. stuff was renormalized after floating point inaccuracies, which cause the roundtrips to fail. Now we store the raw weights instead and normalize on the fly). Still waiting for the cached property support in slotted attrs classes, which will come in the next release (it's already available on main), so same as #74

File: tests/hypothesis_strategies/objectives.py
Comment (AVHopp) [reply tests/hypothesis_strategies/objectives.py:27]: Have a slight tendency to not allowing weights of 0 for the reasons you mention, and also because I do not see a real use case or reason for having weights of 0.

Comment (AVHopp) [reply tests/hypothesis_strategies/objectives.py:27]: However, I think this is currently not being considered in the creation of objectives, right? If I see it correctly, users are able to provide a weight of 0 which is at the very least not consistent with the test here.

File: tests/hypothesis_strategies/targets.py
Comment (AdrianSosic) [reply tests/hypothesis_strategies/targets.py:15]: Yes, it's supposed to be a strategy for having the option to narrow the target strategy (was needed for this PR). Agree, so far the type annotations are not very rigorous in the tests, and I omitted it here simply because it was not obvious to me what the type should be. But simply the fact that you asked shows that we should be a bit more rigorous here, even if it just helps to better understand what's going on. Will try to fix this, perhaps `reveal_type` can help me here. Also, can give a more suggestive parameter name. Leaving this thread open until I find a solution...

Comment (AVHopp) [reply tests/hypothesis_strategies/targets.py:15]: Looking at the import, I would interpret that `intervals` here refers to either a single interval or a colletion of those, since the strategy is probably `st_intervals`. But I fully agree, having more informative type hints would be great and very much appreciated also from me 😄

File: tests/hypothesis_strategies/objectives.py
Comment (AdrianSosic) [reply tests/hypothesis_strategies/objectives.py:27]: Could it be that you simply overlooked the `exclude_min=True`?
BTW: stuff is now validated in the objectives, i.e. it's impossible to provide a zero weight:
image

File: tests/hypothesis_strategies/targets.py
Comment (AVHopp) [inline tests/hypothesis_strategies/targets.py:19]: Why not `if not intervals`? Is this just a question of style? (Just asking out of curiosity)

File: tests/hypothesis_strategies/utils.py
Comment (AVHopp) [inline tests/hypothesis_strategies/utils.py:57]: Looking at the other parts of the code, I would expect that you could also implement this via using `draw(ffloats)` similar to how it is done for the half-bounded intervals. In case this is true, what is the reason to not do it and rely on `hnp` instead?

File: tests/serialization/test_parameter_serialization.py
Comment (AVHopp) [inline tests/serialization/test_parameter_serialization.py:12]: Just to be extra annoying (pls dont get angry 😢 ): This file would not have been part of the PR if it was only dealing with Hypothesis for objective 😜

**Review by AVHopp — COMMENTED** [review]
Looking good in general, but there are still some parts I'd like your opinion on.

File: tests/hypothesis_strategies/objectives.py
Comment (AdrianSosic) [reply tests/hypothesis_strategies/objectives.py:27]: Hi @AVHopp, I think we responded at the exact same moment so I didn't see your message until now. Why do you think users are able to provide 0.0 as weights?

File: baybe/objectives/desirability.py
Comment (AdrianSosic) [reply baybe/objectives/desirability.py:114]: I know it seems weird at first glance but it's actually an elegant (I think) solution to a problem related to subclasses not sharing the same attributes. Don't want to type many sentences here to explain so:
* If you are curious, try to find out yourself why simply using a `targets` attribute would not work
* If you trust me and don't care: resolve
* If you want to now but are lazy: call me 
:D

File: baybe/objectives/desirability.py
Comment (AdrianSosic) [reply baybe/objectives/desirability.py:119]: That's how it used to be but this causes roundtrip problems (mentioned this briefly in a meeting and see also comment to Martin's question here in this PR). The problem is that if you normalize [1, 1, 1], you get floating point inaccuracies, and then if you do a serialization roundtrip, stuff will enter the converter again and subsequently re-normalized, causing equality checks to fail. 

There could be other ways (e.g. somehow avoid that second normalization from happening) but this was the easiest solution I could come up with that doesn't cause any overhead.

File: baybe/objectives/desirability.py
Comment (AdrianSosic) [reply baybe/objectives/desirability.py:147]: But I don't see what's wrong with the comment? The step is literally the transformation into desirability values. Explicitly mentioning something like "normalize weights and ..." would even be bad because we are not calling a function but a property (which is to be interpreted like an attribute), if that is what you mean. So this should be also handled text-wise as if it was an attribute, because the computation is hidden away at this point.

File: tests/hypothesis_strategies/parameters.py
Comment (AdrianSosic) [reply tests/hypothesis_strategies/parameters.py:22]: You mean the renamings? I'll let @Scienfitz decide. If he agrees, will create an intermediate PR containing that one commit (is not much work, indeed)

File: tests/serialization/test_parameter_serialization.py
Comment (AdrianSosic) [reply tests/serialization/test_parameter_serialization.py:12]: Yes, so actually same issue as mentioned in the other thread. Will close this one here so we have only one point of discussion

File: tests/hypothesis_strategies/targets.py
Comment (AdrianSosic) [reply tests/hypothesis_strategies/targets.py:19]: No, rather because there is a difference between `if x` and `if x is not None` or respectively `if not x` and `if x is None`: it's the exact same issue that we had in Tobi's error fix (other things will also evaluate falsy, such as empty lists etc). Here, the situation is even less clear because of the missing type hint, which makes it even unspecified what types `intervals` can even take except for `None`. If we placed something more precise, such as `interval: Optional[Interval] = None`, then it ultimately boils down to the question if objects of type `Interval` can evaluate to false or not.

File: tests/hypothesis_strategies/objectives.py
Comment (Scienfitz) [reply tests/hypothesis_strategies/objectives.py:27]: >> Could it be that you simply overlooked the exclude_min=True?

looked at it twice and still overlooked it :D
Resolved

File: tests/hypothesis_strategies/parameters.py
Comment (Scienfitz) [reply tests/hypothesis_strategies/parameters.py:22]: Let us discuss productive things and not this

File: tests/hypothesis_strategies/utils.py
Comment (AdrianSosic) [reply tests/hypothesis_strategies/utils.py:57]: Yes, was my first attempt, in fact. But it turned out the `hnp` version was much more elegant due to the built-in `unique=True` feature, which is crucially needed here. The alternative would have been quite messy, because you'd need to first create the lower bound and then based on it the upper, but the problem is that while creating the lower you still need to ensure that there is still "space" for the upper. So it would have involved some nasty handling of `np.nextafter` etc. The current solution does it in one atomic step, no rejection sampling involved 😎

File: baybe/objectives/desirability.py
Comment (AVHopp) [reply baybe/objectives/desirability.py:114]: If you tell me that there is a reason behind it I trust you ;) Just wanted to make sure that this is the case as this could also have been an occasion of just overlooking a potential easier solution ;)

File: baybe/objectives/desirability.py
Comment (AVHopp) [reply baybe/objectives/desirability.py:147]: True, I was interpreting the `_normalized_weights` here as a function, which it isn't.

File: baybe/objectives/desirability.py
Comment (AdrianSosic) [reply baybe/objectives/desirability.py:114]: appreciated

**Review by AVHopp — APPROVED** [review]

**Pull request merged**
